client/src/main/java/io/pravega/client/connection/impl/RawClient.java

```diff
@@ -98,23 +98,25 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
 
     public RawClient(PravegaNodeUri uri, ConnectionPool connectionPool) {
         this.segmentId = null;
-        this.connection = connectionPool.getClientConnection(flow, uri, responseProcessor)
-        .exceptionally(e -> {
+        this.connection = new CompletableFuture<>();
+        connectionPool.getClientConnection(flow, uri, responseProcessor, this.connection);
+        this.connection.exceptionally(e -> {
             log.warn("Exception observed while attempting to obtain a connection to segment store {}", uri, e);
             throw new CompletionException(new ConnectionFailedException(e));
         });
-        Futures.exceptionListener(connection, e -> closeConnection(e));
+        Futures.exceptionListener(connection, this::closeConnection);
     }
 
     public RawClient(Controller controller, ConnectionPool connectionPool, Segment segmentId) {
         this.segmentId = segmentId;
-        this.connection = controller.getEndpointForSegment(segmentId.getScopedName())
-                                    .thenCompose((PravegaNodeUri uri) -> connectionPool.getClientConnection(flow, uri, responseProcessor))
+        this.connection = new CompletableFuture<>();
+        controller.getEndpointForSegment(segmentId.getScopedName())
```

Comment:
if controller.getEndpointForSegment(..) complets exceptionally then a ConnectionFailedException will not be thrown...